lightning/src/ln/payment_tests.rs

```diff
@@ -236,6 +236,62 @@ fn mpp_receive_timeout() {
 	do_mpp_receive_timeout(false);
 }
 
+#[test]
+fn test_keysend_payments() {
+	do_test_keysend_payments(false, false);
+	do_test_keysend_payments(false, true);
+	do_test_keysend_payments(true, false);
+	do_test_keysend_payments(true, true);
+}
+
+fn do_test_keysend_payments(public_node: bool, with_retry: bool) {
+	let chanmon_cfgs = create_chanmon_cfgs(2);
+	let node_cfgs = create_node_cfgs(2, &chanmon_cfgs);
+	let node_chanmgrs = create_node_chanmgrs(2, &node_cfgs, &[None, None]);
+	let nodes = create_network(2, &node_cfgs, &node_chanmgrs);
+
+	if public_node {
+		create_announced_chan_between_nodes(&nodes, 0, 1);
+	} else {
+		create_chan_between_nodes(&nodes[0], &nodes[1]);
+	}
+	let payer_pubkey = nodes[0].node.get_our_node_id();
+	let payee_pubkey = nodes[1].node.get_our_node_id();
+	let route_params = RouteParameters {
+		payment_params: PaymentParameters::for_keysend(payee_pubkey, 40, false),
+		final_value_msat: 10000,
+	};
+
+	let network_graph = nodes[0].network_graph.clone();
+	let channels = nodes[0].node.list_usable_channels();
+	let first_hops = channels.iter().collect::<Vec<_>>();
+	let first_hops = if public_node { None } else { Some(first_hops.as_slice()) };
+
+	let scorer = test_utils::TestScorer::new();
+	let random_seed_bytes = chanmon_cfgs[1].keys_manager.get_secure_random_bytes();
+	let route = find_route(
+		&payer_pubkey, &route_params, &network_graph, first_hops,
+		nodes[0].logger, &scorer, &(), &random_seed_bytes
+	).unwrap();
+
+	let test_preimage = PaymentPreimage([42; 32]);
```

Comment:
I think that's okay since `pass_along_path` checks that the preimage from `Event::PaymentClaimable` matches the expected one passed in?

I had mainly noticed that a problem was that we never tested a full payment using `spontaneous_with_retry` which uses `OutboundPayment::send_payment_internal` whereas the method without retries doesn't